### PR TITLE
RFC: Enhance version query to support mixed CEX levels

### PR DESCRIPTION
**Pull request by ifranzki**
To allow configurations where mixed CEX levels are online and APQN_ANY is used or an APQN whitelist is specified with APQNs of different CEX levels included, I have reworked the EP11 version query.

I am now building a list of CEX levels of the configured APQNs (ANY or whitelist). All APQNs of the same  CEX level must still have the same firmware level. APQNs of different CEX levels can have different firmware levels.

Checking if a certain feature is supported by all configured APQNs is slightly more complicated now, since you need to check for the minimum firmware level per CEX level. A helper function is added to do this more easily.

### CODE REVIEW

File: usr/lib/ep11_stdll/ep11_specific.c [hunk 1]
```diff
@@ -7765,77 +7772,91 @@ static CK_RV version_query_handler(uint_32 adapter, uint_32 domain,
...
-       memcpy(qv->serialNumber, xcp_info.serialNumber,
-              sizeof(qv->serialNumber));
-   qv->first = FALSE;
-
-   return CKR_OK;
+        /* card may no longer be online, so ignore this error situation */
+        return CKR_OK;
+    }
+
+    new_card_type = FALSE;
```
[the remark this answers is not in the document]
Comment (ifranzki) [reply usr/lib/ep11_stdll/ep11_specific.c]: fixd

File: usr/lib/ep11_stdll/ep11_specific.c [hunk 2]
```diff
@@ -7765,77 +7772,91 @@ static CK_RV version_query_handler(uint_32 adapter, uint_32 domain,
...
+    card_version = qv->ep11_data->card_versions;
+    while (card_version != NULL) {
+        if (card_version->card_type == card_type)
+           break;
+        card_version = card_version->next;
+    }
+
+    if (card_version == NULL) {
+        card_version = calloc(1, sizeof(ep11_card_version_t));
```
[the remark this answers is not in the document]
Comment (ifranzki) [reply usr/lib/ep11_stdll/ep11_specific.c]: fixed

File: usr/lib/ep11_stdll/ep11_specific.c [hunk 3]
```diff
@@ -7765,77 +7772,91 @@ static CK_RV version_query_handler(uint_32 adapter, uint_32 domain,
...
+        card_version->card_type = card_type;
+        card_version->firmware_API_version = xcp_info.firmwareApi;
+        card_version->firmware_version = xcp_info.firmwareVersion;
+
+        card_version->next = qv->ep11_data->card_versions;
+        qv->ep11_data->card_versions = card_version;
+
+        new_card_type = TRUE;
+    }
+
```
[the remark this answers is not in the document]
Comment (ifranzki) [reply usr/lib/ep11_stdll/ep11_specific.c]: Good point. Fixed.

**Comment by p-steuer** [comment]
Lgtm. As you labeled this PR "RFC", would you like to have a discussion or can this be merged ?

**Comment by ifranzki** [comment]
> Lgtm.  As you labeled this PR "RFC", would you like to have a discussion or can this be merged ?

If we all agree that this is good as it is now, then lets merge it.

**Pull request merged**